### PR TITLE
Advantages of TypeScript #39520

### PR DESCRIPTION
**Pull request by shashankbaranwal**
- learn about typescript
- learn about some basic command
- learn about interface

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+node_modules
+package-lock.json
```
Comment (namanpar) [inline .gitignore:2]: add dist folder

**Review by namanpar — CHANGES_REQUESTED** [review]
First remove the unnecessary files

**Review by namanpar — CHANGES_REQUESTED** [review]
Kindly resolve conflicts and node_modules from pr

File: extra/patterns/index.js [hunk 2]
```diff
@@ -1,4 +1,10 @@
+<<<<<<< HEAD
+import creatediamond from './diamond';
+import createequilatral from './equilateral';
+
+=======
 import creatediamond from './diamond';
 import createequilatral from './equilateral';
 
+>>>>>>> 146c3daaffaa716bb8480bc3e09ba73da5e7895f
```
Comment (namanpar) [inline extra/patterns/index.js:9]: Remove unnecessary/duplicate code

File: extra/utils/Permissions.js [hunk 3]
```diff
@@ -15,3 +15,5 @@ export default function hasPermission(moduleName,role,permissionType)
     }
     return true;
 }
+
+>>>>>>> 146c3daaffaa716bb8480bc3e09ba73da5e7895f
```
Comment (namanpar) [inline extra/utils/Permissions.js:19]: Remove the unnecessary/duplicate code

File: extraTS/utils/Permissions.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        if (type.includes(role))
+            return true
+        else
+            return false;
+    }
+}
+// const { getUsers, getUser } = permissions;
+// let result = hasPermission(getUsers, 'head-trainer', 'Delete');
+// let result_1 = hasPermission(getUsers, 'trainer', 'all');
+// console.log(result);
```
Comment (namanpar) [inline extraTS/utils/Permissions.ts:26]: Create the function as per instruction in task

File: extraTS/utils/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+export default function validateEmail(email){
+    const regex = /\w+.\w+@successive.tech$/i;
+    if(regex.test(email))
+    {
+        return true;
+    }
+    else{
+        return false;
+    }
+}
```
Comment (namanpar) [inline extraTS/utils/helpers.ts:10]: Write the return type of function

File: extraTs/utils/Validation.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import validateEmail from './helpers'
+
+export default function validateUsers(users : IUsers[]) : void{
+    let countValid =0;
+    let countInvalid  =0;
```
Comment (namanpar) [inline extraTs/utils/Validation.ts:5]: Write the return type of these variables

**Review by namanpar — CHANGES_REQUESTED** [review]
Kindly do the requested changes and remove the node_modules for commit

File: extraTs/utils/Permissions.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { permissions } from '../constant';
+export default function hasPermission( moduleName: string , role: string , permissionType: string)
+{
+    console.log(permissions);
+    console.log(moduleName);
```
Comment (namanpar) [inline extraTs/utils/Permissions.ts:5]: Enter return type of function

File: .gitignore (hunk 1 above)
Comment (shashankbaranwal) [reply .gitignore]: Added dist folder

File: extra/patterns/index.js (hunk 2 above)
Comment (shashankbaranwal) [reply extra/patterns/index.js:9]: I removed all the unnecessary file and duplicate code which is present in that files.

File: extra/utils/Permissions.js (hunk 3 above)
Comment (shashankbaranwal) [reply extra/utils/Permissions.js:19]: I removed all the unnecessary file and duplicate code which is present in that files.

File: extraTS/utils/Permissions.ts (hunk 4 above)
Comment (shashankbaranwal) [reply extraTS/utils/Permissions.ts]: I make the function as per given in the task but by mistaken make wrong the folder. I resolved it

File: extraTS/utils/helpers.ts (hunk 5 above)
Comment (shashankbaranwal) [reply extraTS/utils/helpers.ts]: I make the function as per given in the task but by mistaken make wrong the folder .i resolved it.

File: extraTs/utils/Validation.ts (hunk 6 above)
Comment (shashankbaranwal) [reply extraTs/utils/Validation.ts]: i forgot to write the return type of this variable.

File: extraTs/utils/Permissions.ts (hunk 7 above)
Comment (shashankbaranwal) [reply extraTs/utils/Permissions.ts]: I forgot to write the return type of function in this files.

**Review by namanpar — CHANGES_REQUESTED** [review]
Still two issues remain. Kindly resolve them

**Review by namanpar — APPROVED** [review]

**Pull request merged**